### ISSUE TITLE
(*http.Request).Post() panics when passed nil body in 1.6.2

### ISSUE THREAD

**Issue by erikh** [golang/go#15455]
1. What version of Go are you using (`go version`)?

1.6.2

2. What operating system and processor architecture are you using (`go env`)?

I tested most of this on the play.golang.org site -- links are provided below. 

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

The problem is that when passing nil body to `(*http.Request).Post()` the code type switches on the various types it supports. However, it does not actually check for a straight nil (passing `io.Reader(nil)` works just fine) because the interface obscures it.

The second request here will panic.

https://play.golang.org/p/kaLpNTYIEm

The issue is here:

https://golang.org/src/net/http/request.go?s=19631:19703#L612

And an example of (what I think is) an appropriate fix is here: https://play.golang.org/p/yaFaol2nyl

4. What did you expect to see?

no panics :)

5. What did you see instead?

panics.

Final notes: <3. I'm sorry to pitch this over the fence, but I am really strapped for time.

**Comment by OneOfOne** [comment]
Why in the world would you do that and why do you expect it to work?

**Comment by erikh** [comment]
Probably because the docs say it accepts an optional body, there is a 
ton of code to guard against nils in there, and it panics in the middle 
of a library.

https://golang.org/pkg/net/http/#NewRequest

My use-case is kind of irrelevant.

**Comment by griesemer** [comment]
@erikh Your body is not nil. You are passing in a _concrete_ *bytes.Buffer that happens to have a nil value, but when assigned to the argument, an io.Reader _interface_, the value is a non-nil interface and thus internal nil checks can't catch it. This is working as expected.

**Comment by erikh** [comment]
ok.

**Comment by erikh** [comment]
perhaps you should correct the documentation instead then.

**Comment by OneOfOne** [comment]
@erikh the documentation is correct, you're misunderstanding how interfaces work, check this example:

https://play.golang.org/p/DB3OR7duMB

**Comment by griesemer** [comment]
One could guard against it by rewriting the code in NewRequest
```
if body != nil {
	switch v := body.(type) {
	case *bytes.Buffer:
		req.ContentLength = int64(v.Len())
	case *bytes.Reader:
		req.ContentLength = int64(v.Len())
	case *strings.Reader:
		req.ContentLength = int64(v.Len())
	}
}
```
as
```
	switch v := body.(type) {
        case nil:
           // nothing to do
	case *bytes.Buffer:
                if v != nil {
                    req.ContentLength = int64(v.Len())
                }
	case *bytes.Reader:
                if v != nil {
		     req.ContentLength = int64(v.Len())
                }
	case *strings.Reader:
                if v != nil {
		     req.ContentLength = int64(v.Len())
                }
	}
```
But this tends to be overkill and it's not like the rest our std library is written.

**Comment by erikh** [comment]
Nevermind guys. Thanks for the responses.

**Comment by erikh** [comment]
Right, that's the exact fix I submitted in example.

**Comment by bradfitz** [comment]
There's no need for any changes. This is a common mistake when learning Go. It's even in the FAQ: https://golang.org/doc/faq#nil_error

**Comment by erikh** [comment]
I'm not learning go. If you see the example, it *clearly* demonstrates that I fully comprehend the problem, I just figured it was *smart* to avoid *panicking* in a *core library*.

I have pedigree you guys are familiar with in the golang community, but I'm not just going to sell myself to you to stop being explained golang like i'm 5.

This whole conversation is ridiculously condescending. I'm going to unsubscribe now. Read the links next time.

**Comment by bradfitz** [comment]
@erikh, no, I did see your example. That's why I replied with the entry to the FAQ. The docs say the body can be nil, but the body argument to NewRequest is an interface, so that means it has to be a nil interface value. You put a nil concrete type (*bytes.Buffer) into an interface value on line 14 of your https://play.golang.org/p/kaLpNTYIEm (where you commented `// panics!! but why?`), resulting in a non-nil interface value. This is where you violated the contract of the documentation.

I haven't seen you around the Go community before, so I assumed you were still learning Go. My apologies if that's not the case. In any case, we're all still learning. After 6+ years with Go, I still find new things about it.
